test/fs.js

```diff
@@ -0,0 +1,171 @@
+'use strict';
+
+const metatests = require('metatests');
+const metasync = require('metasync');
+const common = require('..');
+const fs = require('fs');
+const path = require('path');
+const stream = require('stream');
+
+const testData =
+  'JySAF1figbONWxAeeoGhDXeI7df25GKXKPcyCl63ZjIv5HZGQH8yfFeHGGaSJ9zDMP0CwtgX2madIQkFmuq38SNhcKcGQ';
+const absoluteLink = __dirname.replace('test', 'lib');
+const srcFiles = [
+  'srcDir',
+  'srcDir/dir1',
+  'srcDir/dir2',
+  'srcDir/file1',
+  'srcDir/link',
+  'srcDir/dir1/subdir1',
+];
+
+const dstFiles = [
+  'dstDir',
+  'dstDir/dir1',
+  'dstDir/dir2',
+  'dstDir/file1',
+  'dstDir/link',
+  'dstDir/dir1/subdir1',
+];
+
+class ToUpperCase extends stream.Transform {
+  constructor(options = {}) {
+    options.decodeStrings = false;
+    super(options);
+  }
+  _transform(chunk, encoding, callback) {
+    if (encoding !== 'utf8') {
+      this.emit('error', new Error('only utf8'));
+      return callback();
+    }
+    this.push(chunk.toUpperCase());
+    return callback();
+  }
+}
+
+const transOptions = {
+  Transform: ToUpperCase,
+  readOptions: { encoding: 'utf-8' },
+  writeOptions: {},
+};
+
+const rmdirp = (dir, cb) => {
+  fs.lstat(dir, (err, stats) => {
+    if (err) {
+      if (err.code === 'ENOENT') cb(null);
+      else cb(err);
+      return;
+    }
+    if (stats.isDirectory()) {
+      fs.readdir(dir, (err, files) => {
+        if (err) {
+          cb(err);
+          return;
+        }
+        files = files.map(file => path.join(dir, file));
+        metasync.each(files, rmdirp, err => {
+          if (err) cb(err);
+          else fs.rmdir(dir, cb);
+        });
+      });
+    } else {
+      fs.unlink(dir, cb);
+    }
+  });
+};
+
+const copydirTest = metatests.test('Recursive dir copy');
+copydirTest.endAfterSubtests();
+
+copydirTest.beforeEach((test, cb) => {
+  metasync.series(
+    srcFiles,
+    (file, cb) => {
+      if (path.basename(file).startsWith('file')) {
+        fs.appendFile(file, testData, cb);
+      } else if (file.includes('link')) {
```

Comment:
```suggestion
      } else if (path.basename(file).startsWith('link')) {
```

Reply:
Just explain me, why do you do it this way?

Reply:
This is related to https://github.com/metarhia/common/pull/251#discussion_r260271715. For example, this should be valid.
```js
const srcFiles = [
  'linkDir1',
  'srcDir',
  'srcDir/link1', // it will be linked to `linkDir1`,
  'srcDir/link1/file1', // it also will be linked to `linkDir1`,
];
```
It will make directory `linkDir1`, create a symlink from `srcDir/link1` to it and create a file `file1`. And though it doesn't make much sense in this context as it will be much better to have something like this.
```js
const srcFiles = [
  'linkDir1',
  'linkDir1/file1',
  'srcDir',
  'srcDir/link1', // it will be linked to `linkDir1`,
];
```
Also, `path.basename` and `String.prototype.startsWith` should work faster than `String.prototype.includes` if you put `path.basename(file)` into a separate variable though v8 might optimise it. You can learn more about it [here](https://github.com/nodejs/node/blob/master/lib/path.js), [here](https://chromium.googlesource.com/v8/v8/+/refs/heads/7.0.276/src/builtins/builtins-string.cc) and [here](https://chromium.googlesource.com/v8/v8/+/refs/heads/7.0.276/src/builtins/builtins-string-gen.cc).
Another thing is code style consistency: having similar conditions will make it easier to read.

Reply:
And a few more things about links. Now if `dereference` option is true and there are recursive symlinks `copyDir` will fail.
```js
const srcFiles = [
  'srcDir/dir1',
  'srcDir/dir1/link2', // linked to `dir1`,
// or
  'srcDir/dir2',
  'srcDir/dir3',
  'srcDir/dir2/link3', // linked to `dir3`,
  'srcDir/dir3/link4', // linked to `dir2`,
];
```
This is something that we don't need right now, but if you feel about it you can try to do this.